From: Coly Li <colyli@suse.de>
Author: Coly Li <colyli@suse.de>
Date:	Thu Apr 14 16:31:02 CST 2016
Subject: [PATCH] Add wait_event_cmd()
Patch-mainline: never, it is a modified version for SLE12-SP1 kernel
Reference: bsc#953048

This patch is a modified version of upstream linux commit 82e06c81 (wait: 
add wait_event_cmd()), I use code infrastructure from SLE12-SP1 kernel to
rewrite this patch.

Signed-off-by: Coly Li <colyli@suse.de>
---
 include/linux/wait.h |   29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

--- a/include/linux/wait.h
+++ b/include/linux/wait.h
@@ -218,6 +218,35 @@ do {									\
 	__wait_event(wq, condition);					\
 } while (0)
 
+/**
+ * wait_event_cmd - sleep until a condition gets true
+ * @wq: the waitqueue to wait on
+ * @condition: a C expression for the event to wait for
+ * cmd1: the command will be executed before sleep
+ * cmd2: the command will be executed after sleep
+ *
+ * The process is put to sleep (TASK_UNINTERRUPTIBLE) until the
+ * @condition evaluates to true. The @condition is checked each time
+ * the waitqueue @wq is woken up.
+ *
+ * wake_up() has to be called after changing any variable that could
+ * change the result of the wait condition.
+ */
+#define wait_event_cmd(wq, condition, cmd1, cmd2)			\
+do {									\
+	DEFINE_WAIT(__wait);						\
+									\
+	for (;;) {							\
+		prepare_to_wait(&wq, &__wait, TASK_UNINTERRUPTIBLE);	\
+		if (condition)						\
+			break;						\
+		cmd1;							\
+		schedule();						\
+		cmd2;							\
+	}								\
+	finish_wait(&wq, &__wait);					\
+} while(0)
+
 #define __wait_event_timeout(wq, condition, ret)			\
 do {									\
 	DEFINE_WAIT(__wait);						\
